### PR TITLE
containers as functions

### PR DESCRIPTION
**Pull request by sahil-lakhwani**
Signed-off-by: sahil-lakhwani <sahilakhwani@gmail.com>

Addresses #235 

This PR adds a feature to support Docker containers as functions. User will be able to specify a Docker image which will be run as a Fission Function.
It assumes the Docker image exposes port 8888 to serve any HTTP requests.

Changes: 
CLI: A new parameter `--image` is added for `fn create` and `fn update`
Executor: A new executor `container` is added which creates the necessary Kubernetes resources.

Example: 
`fission fn create  --executortype=container --name=carts --port=80 --command="./app" --args="--port=80"   --image weaveworksdemos/catalogue:0.3.5
`

---
This change is [Reviewable](https://reviewable.io/reviews/fission/fission/1681)

### CODE REVIEW

**Comment by vishal-biyani** [comment]
Hey @sahil-lakhwani A few things that would be good before we can start 

1. Can you please add a working example of a container function, which should be different from usual functions.
2. An ideal example would be a microservice with more than a few API endpoints and should be possible to access without the need for creating multiple routing endpoints
3. Also some tests - unit, as well as integration, would be really good.

File: pkg/executor/executortype/container/container.go [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+	service := &apiv1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        svcName,
+			Labels:      deployLabels,
+			Annotations: deployAnnotations,
+		},
+		Spec: apiv1.ServiceSpec{
+			Ports: []apiv1.ServicePort{
+				{
+					Name:       "http-env",
```
Comment (harshthakur9030) [inline pkg/executor/executortype/container/container.go:324]: Hey @sahil-lakhwani , are the values here intentionally hard-coded?

**Comment by RealHarshThakur** [comment]
Some of the pending items which can be addressed in other PRs are:

- Client-side redirects: Supporting redirects based on user input. As of now, frontend applications may try to fetch JS and CSS files at `/` when the function is hosted at `/frontend` . The list of redirects could get exhaustive as a lot of such files can be used.

- Scaling function pods to 0 by enabling HPA based on metrics from Router. Simplest way to achieve this would be to use KEDA. If not, we might have to install other third party component like [this](https://github.com/DirectXMan12/k8s-prometheus-adapter). This also applies to NewDeploy

File: pkg/executor/executortype/container/common.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	k8s_err "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+
+	fv1 "github.com/fission/fission/pkg/apis/core/v1"
+)
+
+// getResources gets the resources(CPU, memory) set for the function
+func (cn *Container) getResources(fn *fv1.Function) apiv1.ResourceRequirements {
```
Comment (vishal-biyani) [inline pkg/executor/executortype/container/common.go:35]: nitpick - wonder if we can modify the similar method here [func (deploy *NewDeploy) getResources(env *fv1.Environment, fn *fv1.Function) apiv1.ResourceRequirements {](https://github.com/fission/fission/blob/04350fc0d7c18baa23158388fc22d989cd61aa30/pkg/executor/executortype/newdeploy/newdeploy.go#L337-L338) to reuse common logic. We can also refactor later, not going to hold this PR for just this

**Review by vishal-biyani — APPROVED** [review]
Largely looks good to me, one small nitpick comment - we can address in this one or next one.

File: pkg/executor/executortype/container/common.go (hunk 2 above)
Comment (sanketsudake) [reply pkg/executor/executortype/container/common.go:35]: Hi Vishal, we will fix it in another cycle. There are lot of places where we need to combine code across executors.

File: pkg/router/functionHandler.go [hunk 3]
```diff
@@ -257,6 +258,9 @@ func (roundTripper *RetryingRoundTripper) RoundTrip(req *http.Request) (*http.Re
 				req.URL.Path = "/"
 			}
 
+			roundTripper.logger.Info(req.URL.Host)
```
Comment (sanketsudake) [inline pkg/router/functionHandler.go:261]: Remove these statements and unwanted code.

File: pkg/utils/utils.go [hunk 4]
```diff
@@ -47,6 +47,20 @@ func UrlForFunction(name, namespace string) string {
 	return fmt.Sprintf("%v/%v", prefix, name)
 }
 
+// RemoveFuncName removes the name of the function
```
Comment (sanketsudake) [inline pkg/utils/utils.go:50]: Probably not required, remove after evaluation.

**Pull request merged**